packages/dashboard/src/components/alert-store.tsx

```diff
@@ -31,42 +17,71 @@ enum AlertCategory {
 export const AlertStore = React.memo(() => {
   const rmf = React.useContext(RmfAppContext);
   const [taskAlerts, setTaskAlerts] = React.useState<Record<string, Alert>>({});
-  const [robotAlerts, setRobotAlerts] = React.useState<Record<string, Alert>>({});
+  const refreshAlertCount = React.useRef(0);
+
+  const categorizeAndPushAlerts = (alert: Alert) => {
+    // We check if an existing alert has been acknowledged, remove it before
+    // adding the acknowledged alert.
+    switch (alert.category) {
+      case AlertCategory.Task:
+        setTaskAlerts((prev) => {
+          const filteredTaskAlerts: Record<string, Alert> = {};
+          for (let key in prev) {
+            if (key !== alert.original_id) {
+              filteredTaskAlerts[key] = prev[key];
+            }
+          }
+          filteredTaskAlerts[alert.id] = alert;
+          return filteredTaskAlerts;
+        });
+        break;
+      default:
+      // do nothing in default case
+    }
+  };
+
+  React.useEffect(() => {
+    const subs: Subscription[] = [];
+    subs.push(
+      AppEvents.alertListOpenedAlert.subscribe((alert) => {
+        if (alert) {
+          categorizeAndPushAlerts(alert);
+        }
+      }),
+    );
+    return () => subs.forEach((s) => s.unsubscribe());
+  }, []);
 
   React.useEffect(() => {
     if (!rmf) {
       return;
     }
     const sub = rmf.alertObsStore.subscribe(async (alert) => {
-      switch (alert.category) {
-        case AlertCategory.Task:
-          setTaskAlerts((prev) => ({ ...prev, [alert.id]: alert }));
-          break;
-        case AlertCategory.Fleet:
-        case AlertCategory.Robot:
-          setRobotAlerts((prev) => ({ ...prev, [alert.id]: alert }));
-          break;
-        default:
-        // do nothing in default case
-      }
+      categorizeAndPushAlerts(alert);
+      refreshAlertCount.current += 1;
+      AppEvents.refreshAlertCount.next(refreshAlertCount.current);
     });
     return () => sub.unsubscribe();
   }, [rmf]);
 
-  const taskAlertKey: Partial<typeof taskAlerts> = { ...taskAlerts };
   const removeTaskAlert = (id: string) => {
```

Comment:
although we don't expect to handle many task alerts at the same time, I would prefer to use a traditional `for ... in` loop here for performance, since there will be 1 less lookup for deletion, and it looks like a spread operator is generally less performant, https://www.measurethat.net/Benchmarks/Show/16078/0/javascript-spread-operator-vs-objectassign-vs-for-in-lo#latest_results_block

Reply:
I did a bit more digging and found that the general consensus of typescript prioritizes readability and async operations, I've refactored it to use `Object.entries` after seeing some of your examples, https://github.com/open-rmf/rmf-web/pull/710/commits/f843fb92cd35e06af5fe64585a298becbe6f02fd